File: Dynamic Programming/Medium/LET_no764_Largest Plus Sign.cpp
class Solution {
public:
    int orderOfLargestPlusSign(int n, vector<vector<int>>& mines) {
        vector<vector<int>> dp(n, vector<int>(n, n));
        for (auto& m : mines)
            dp[m[0]][m[1]] = 0;
        
        for (int i = 0; i < n; i++) {
            for (int j = 0, k = n - 1, left = 0, right = 0, up = 0, down = 0; j < n; j++, k--) {
                dp[i][j] = min(dp[i][j], left = (dp[i][j] ? left + 1 : 0));
                dp[i][k] = min(dp[i][k], right = (dp[i][k] ? right + 1 : 0));
                dp[j][i] = min(dp[j][i], up = (dp[j][i] ? up + 1 : 0));
                dp[k][i] = min(dp[k][i], down = (dp[k][i] ? down + 1 : 0));
            }
        }
        
        int res = 0;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++)
                res = max(res, dp[i][j]);
        }
        return res;
    }
};

File: Dynamic Programming/Medium/LET_no764_Largest Plus Sign.py
class Solution:
    def orderOfLargestPlusSign(self, n: int, mines: List[List[int]]) -> int:
        dp = [[n] * n for _ in range(n)]
        for i, j in mines:
            dp[i][j] = 0
        
        for i in range(n):
            left = right = up = down = 0
            for j, k in zip(range(n), reversed(range(n))):
                left = left + 1 if dp[i][j] else 0
                dp[i][j] = min(dp[i][j], left)
                right = right + 1 if dp[i][k] else 0
                dp[i][k] = min(dp[i][k], right)
                up = up + 1 if dp[j][i] else 0
                dp[j][i] = min(dp[j][i], up)
                down = down + 1 if dp[k][i] else 0
                dp[k][i] = min(dp[k][i], down)
        
        return max(map(max, dp))